constraint-solver/src/quadratic_symbolic_expression.rs

```diff
@@ -403,6 +427,53 @@ impl<T: FieldElement, V: Ord + Clone + Hash + Eq + Display> QuadraticSymbolicExp
         };
         Some(Effect::RangeConstraint(solve_for.clone(), constraint))
     }
+
+    /// Tries to detect a range constraint of the form `(X - value1) * (X - value2) = 0`.
+    /// The most common example would be a bit constraint.
+    /// In that case, we can return a range constraint for `X` and mark the
+    /// constraint as complete.
+    // TODO: This function would not detect some equivalent constraints, like
+    // `X * X - X`.
+    // This could be fixed by finding a canonical form for the quadratic
+    // expression, and normalizing the constants.
+    fn try_detect_range_constraint(&self) -> Option<ProcessResult<T, V>> {
+        let (factor1, factor2) = &self.try_as_single_product()?;
+        let (var1, value1) = factor1.try_solve_constant_assignment()?;
+        let (var2, value2) = factor2.try_solve_constant_assignment()?;
+
+        if var1 == var2 {
+            let range_constraint = RangeConstraint::from_value(value1)
+                .disjunction(&RangeConstraint::from_value(value2));
+
+            // If the range width is at most two, the range constraint is equivalent to the polynomial constraint.
+            // Otherwise, we might lose information.
+            let complete = range_constraint.range_width() <= 2.into();
+            let effects = vec![Effect::RangeConstraint(var1, range_constraint)];
+
+            Some(ProcessResult { effects, complete })
+        } else {
+            None
+        }
+    }
+
+    /// Tries to solve a constant assignment of the form `X = a`.
+    fn try_solve_constant_assignment(&self) -> Option<(V, T)> {
+        if self.is_quadratic() || self.linear.len() != 1 {
+            return None;
+        }
+        let ProcessResult {
+            effects,
+            complete: true,
+        } = self.solve(&NoRangeConstraints).ok()?
+        else {
+            return None;
+        };
+        if let [Effect::Assignment(var, expr)] = effects.as_slice() {
+            Some((var.clone(), expr.try_to_number()?))
```

Comment:
Actually we don't really need to know that it is a number.

But yeah, let's merge it.